proposals/0388-bls12-381-syscalls.md

```diff
@@ -0,0 +1,181 @@
+support more advanced ZK or cryptography applications, but these are outside the
+scope of this SIMD for now.
+
+### Addition and Scalar Multiplication on G1
+
+There is already a dedicated `sol_curve_group_op` syscall function for general
+elliptic curve group operations. This function takes in a `curve_id`, `group_op`,
+and two scalar/points that are encoded in little-endian. It interprets the
+curve points according to the curve id, and applies the group operations
+specified by the `group_op`. Currently, the syscall supports the curve25519
+edwards and ristretto representations.
+
+This function can be extended to support the addition, subtraction, and scalar
+multiplication in BLS12-381 G1. We propose adding new `curve_id` constants for
+BLS12-381. The BLS12-381 inputs should be interpreted as affine points unlike
+how Curve25519 points are currently interpreted by the syscall.
+
+```rust
+pub const CURVE25519_EDWARDS: u64 = 0;
+pub const CURVE25519_RISTRETTO: u64 = 1;
+
+// New Curve ID
+pub const BLS12_381: u64 = 2;
+pub const BLS12_381_G1: u64 = 3;
+pub const BLS12_381_G2: u64 = 3;
+
+pub const ADD: u64 = 0;
+pub const SUB: u64 = 1;
+pub const MUL: u64 = 2;
+```
+
+The `BLS12_381_G1` and `BLS12_381_G2` constants will be used for group
+operations and decompression on the respective groups. The `BLS12_381` constant
+will be used for the pairing operations which invovles both.
+
+### Pairing Operation
+
+There is an existing definition for a `sol_curve_pairing_map` syscall function for
+elliptic curve pairing operations.
+
+```rust
+define_syscall!(fn sol_curve_pairing_map(
+    curve_id: u64,
+    point: *const u8,
+    result: *mut u8
+) -> u64);
+```
+
+This function is not actually instantiated at the moment. We propose updating
+this function's signature to take in an array of points in G1 and an array of
+pointst in G2 to support batch pairings.
+
+```rust
+define_syscall!(fn sol_curve_pairing_map(
+    curve_id: u64,
+    g1_points: *const u8,
+    g2_points: *const u8,
+    result: *mut u8
+) -> u64);
+```
+
+The function would interpret `g1_points` as an array of BLS12-381 points in G1
+`[P1, ..., Pn]` and `g2_points` as an array of BLS12-381 points in G2
+`[Q1, ..., Qn]`. Both inputs are interpreted as affine representations
+encoded in little-endian. It should then compute the pairing product
+`e(P1, Q1) * ... * e(Pn, Qn)`.
+
+### Decompression Operations in G1 and G2
+
+For the decompression operations in G1 and G2, we propose adding a dedicated
+syscall for general decompression `sol_curve_decompress`.
+
+```rust
+define_syscall!(fn sol_curve_pairing_map(
+    curve_id: u64,
+    point: *const u8,
+    result: *mut u8
+) -> u64);
+```
+
+This function will take in a curve id and a point that is represented in its
+compressed representation (encoded in little-endian). It will decompress the
+compressed point into an affine representation and write the result.
```

Comment:
Hmm, do you mean, if I have a compressed pubkey and compressed signature, I need to do three syscalls to verify proof of possession? Would that be more expensive than one syscall?